[torchgen] Introduce Executorch types and signatures

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89595
* #89594
* #89487

## Forked `BaseCppType`

Created a module for Executorch: `torchgen.executorch`.

In `torchgen.executorch.api.types.types`:
* Define `BaseCppType` with `torch::executor` namespace.

In `torchgen.executorch.api.et_cpp`:
* Help generate `NamedCType` for `ExecutorchCppSignature` arguments.

In `torchgen.executorch.api.types.signatures`:
* Define the signature using these types. (`ExecutorchCppSignature`)

In `torchgen.executorch.api.types.__init__`:
* Suppress flake8 error for `import *`.

Differential Revision: [D41501836](https://our.internmc.facebook.com/intern/diff/D41501836/)

larryliu0820 committed:
[torchgen] Introduce Executorch types and signatures



Differential Revision: [D41501836](https://our.internmc.facebook.com/intern/diff/D41501836/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Introduce Executorch types and signatures"


Differential Revision: [D41501836](https://our.internmc.facebook.com/intern/diff/D41501836/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Introduce Executorch types and signatures"


## Forked `BaseCppType`

Created a module for Executorch: `torchgen.executorch`.

In `torchgen.executorch.api.types`:
* Define `BaseCppType` with `torch::executor` namespace.
* The signature using these types. (`CppSignature`)

In `torchgen.executorch.api.cpp`:
* Help generate `NamedCType` for `CppSignature` arguments.


Differential Revision: [D41501836](https://our.internmc.facebook.com/intern/diff/D41501836/)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41501836).

larryliu0820 committed:
Update on "[torchgen] Introduce Executorch types and signatures"


## Forked `BaseCppType`

Created a module for Executorch: `torchgen.executorch`.

In `torchgen.executorch.api.types`:
* Define `BaseCppType` with `torch::executor` namespace.
* The signature using these types. (`CppSignature`)

In `torchgen.executorch.api.cpp`:
* Help generate `NamedCType` for `CppSignature` arguments.


Differential Revision: [D41501836](https://our.internmc.facebook.com/intern/diff/D41501836/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Introduce Executorch types and signatures"


## Forked `BaseCppType`

Created a module for Executorch: `torchgen.executorch`.

In `torchgen.executorch.api.types`:
* Define `BaseCppType` with `torch::executor` namespace.
* The signature using these types. (`CppSignature`)

In `torchgen.executorch.api.cpp`:
* Help generate `NamedCType` for `CppSignature` arguments.


Differential Revision: [D41501836](https://our.internmc.facebook.com/intern/diff/D41501836/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Introduce Executorch types and signatures"


## Forked `BaseCppType`

Created a module for Executorch: `torchgen.executorch`.

In `torchgen.executorch.api.types`:
* Define `BaseCppType` with `torch::executor` namespace.
* The signature using these types. (`CppSignature`)

In `torchgen.executorch.api.cpp`:
* Help generate `NamedCType` for `CppSignature` arguments.


Differential Revision: [D41501836](https://our.internmc.facebook.com/intern/diff/D41501836/)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Introduce Executorch types and signatures"


## Forked `BaseCppType`

Created a module for Executorch: `torchgen.executorch`.

In `torchgen.executorch.api.types`:
* Define `BaseCppType` with `torch::executor` namespace.
* The signature using these types. (`CppSignature`)

In `torchgen.executorch.api.cpp`:
* Help generate `NamedCType` for `CppSignature` arguments.


Differential Revision: [D41501836](https://our.internmc.facebook.com/intern/diff/D41501836/)

[ghstack-poisoned]

ezyang commented on torchgen/executorch/api/cpp.py:
```diff
@@ -0,0 +1,409 @@
+#     types.CppSignature, so that we can do stuff like ATen mode.
+#
+#   - defaulting lives here (in fact, the dispatcher is completely
+#     oblivious of defaults!)
```
Instead of copy pasting the old comment, please talk about how this differs from the regular cpp api

ezyang commented on torchgen/executorch/api/cpp.py:
```diff
@@ -0,0 +1,409 @@
+        else:
+            name += "_out"
+
+    return name
```
This should be deduped, I think.

ezyang commented on torchgen/executorch/api/cpp.py:
```diff
@@ -0,0 +1,409 @@
+    if isinstance(t, BaseType):
+        if t.name == BaseTy.Tensor or t.name == BaseTy.Scalar:
+            return None
+        elif str(t) == "SymInt":
```
this is worth a comment

ezyang commented on torchgen/executorch/api/cpp.py:
```diff
@@ -0,0 +1,409 @@
+        return r
+    if isinstance(t, BaseType):
+        if t.name == BaseTy.Tensor:
+            if mutable and not local.use_const_ref_for_mutable_tensors():
```
is executorch able to avoid the mistake we made about mutable Tensor& input? If so, I would much prefer if this knob could be removed entirely.

The mistake here, is passing `Tensor& out`. It is NEVER correct to directly assign over the mutable reference, this is a big footgun in eager PyTorch.

ezyang commented on torchgen/executorch/api/types/types.py:
```diff
@@ -0,0 +1,57 @@
+scalarT = BaseCppType("torch::executor", "Scalar")
+memoryFormatT = BaseCppType("torch::executor", "MemoryFormat")
+intArrayRefT = BaseCppType("torch::executor", "IntArrayRef")
+optionalT = BaseCppType("torch::executor", "optional")
```
Surprised you guys aren't just using `std::optional`

ezyang commented on torchgen/executorch/api/types/signatures.py:
```diff
@@ -0,0 +1,63 @@
+        )
+
+
+from torchgen.executorch.api import cpp
```
don't call this cpp. Call it executorch.

ezyang commented on torchgen/executorch/api/types/signatures.py:
```diff
@@ -0,0 +1,63 @@
+
+
+@dataclass(frozen=True)
+class CppSignature:
```
Don't reuse the same name, it makes it harder to grep for things. ExecutorchCppSignature.

ezyang commented on torchgen/executorch/api/types/signatures.py:
```diff
@@ -0,0 +1,63 @@
+        args_str = ", ".join(args)
+        if name is None:
+            name = self.name()
+        return f"{self.returns_type().cpp_type()} {name}({args_str})"
```
some of the code here, like the string wrangling, is indifferent to your choice of cpp/executorch. It would be nice to deduplicate it.

ezyang commented on torchgen/executorch/api/cpp.py:
```diff
@@ -0,0 +1,409 @@
+        elif isinstance(t.elem, ListType) and str(t.elem.elem) == "int":
+            raise NotImplementedError("Need to implement type resolution for int[]?")
+        elif isinstance(t.elem, ListType) and str(t.elem.elem) == "SymInt":
+            raise NotImplementedError("Need to implement type resolution for SymInt[]?")
```
You don't actually want to implement SymInt directly right, this should devolve to int.

ezyang commented on torchgen/executorch/api/cpp.py:
```diff
@@ -0,0 +1,409 @@
+        elem = argumenttype_type(t.elem, mutable=mutable, binds=binds)
+        return NamedCType(binds, OptionalCType(elem.type))
+    elif isinstance(t, ListType):
+        # TODO: remove these special cases, ArrayRef fallthrough works fine
```
Can Executorch remove these special cases?

ezyang commented on torchgen/executorch/api/cpp.py:
```diff
@@ -0,0 +1,409 @@
+        elif str(t.elem) == "Scalar":
+            return NamedCType(binds, ArrayRefCType(BaseCType(scalarT)))
+        elif str(t.elem) == "Dimname":
+            raise NotImplementedError("Need to implement type resolution for Dimname")
```
no, you just don't want to support dimname in executorch at all lol

ezyang commented on torchgen/executorch/api/cpp.py:
```diff
@@ -0,0 +1,409 @@
+        else:
+            name = fallback_name if len(f.func.returns) == 1 else f"{fallback_name}{i}"
+        returns.append(name)
+    return returns
```
This should be deduped with cpp, I think

larryliu0820 committed:
Update on "[torchgen] Introduce Executorch types and signatures"


## Forked `BaseCppType`

Created a module for Executorch: `torchgen.executorch`.

In `torchgen.executorch.api.types.types`:
* Define `BaseCppType` with `torch::executor` namespace.

In `torchgen.executorch.api.cpp`:
* Help generate `NamedCType` for `CppSignature` arguments.

In `torchgen.executorch.api.types.signatures`:
* Define the signature using these types. (`CppSignature`)

In `torchgen.executorch.api.types.__init__`:
* List out all exported types from `signatures` and `types`. (to make `mypy` happy)

Differential Revision: [D41501836](https://our.internmc.facebook.com/intern/diff/D41501836/)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41501836).

larryliu0820 committed:
Update on "[torchgen] Introduce Executorch types and signatures"


## Forked `BaseCppType`

Created a module for Executorch: `torchgen.executorch`.

In `torchgen.executorch.api.types.types`:
* Define `BaseCppType` with `torch::executor` namespace.

In `torchgen.executorch.api.et_cpp`:
* Help generate `NamedCType` for `ExecutorchCppSignature` arguments.

In `torchgen.executorch.api.types.signatures`:
* Define the signature using these types. (`ExecutorchCppSignature`)

In `torchgen.executorch.api.types.__init__`:
* Suppress flake8 error for `import *`.

Differential Revision: [D41501836](https://our.internmc.facebook.com/intern/diff/D41501836/)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41501836).

larryliu0820 commented:
@pytorchbot merge
